✨ Added redeemed offers filtering for members

ronaldlangeveld opened the pull request:
Adds the ability to filter members based on the offer(s) they redeemed when subscribing.

ref https://github.com/TryGhost/Team/issues/2011

- [x] There's a clear use-case for this code change, explained below
- [x] Commit message has a short title & references relevant issues
- [x] The build will pass (run `yarn test:all` and `yarn lint`)

ronaldlangeveld committed:
Added offer redemption filter model

ref https://github.com/TryGhost/Team/issues/2011

- Added the ability to filter members by redeemed offers via API.

ronaldlangeveld committed:
Added offers filtering frontend

ref https://github.com/TryGhost/Team/issues/2011

- added filtering logic to allow filtering members based on offers
  redeemed.
- wip

ronaldlangeveld committed:
Added multiple select

ref https://github.com/TryGhost/Team/issues/2011

ronaldlangeveld committed:
Merge branch 'main' into tier-filters

ronaldlangeveld committed:
Merge branch 'main' into tier-filters

ronaldlangeveld committed:
Merge branch 'TryGhost:main' into tier-filters

ronaldlangeveld committed:
Removed `offer_redemptions` from relationships

ref https://github.com/TryGhost/Team/issues/2011

- not needed, adds additional info to certain requests where its not
  relevant.

ronaldlangeveld committed:
Added filter test

ronaldlangeveld committed:
Added redemption count link to filtered members

ref https://github.com/TryGhost/Team/issues/2011

- Redemption count in the offers tab now links to the filtered members
  page.

ronaldlangeveld committed:
Fixed linting errors

ronaldlangeveld committed:
Merge branch 'main' into tier-filters

ronaldlangeveld committed:
Added test for members model filter

ref https://github.com/TryGhost/Team/issues/2011

- Added test to ensure the new offer redemption relation filter works.

ronaldlangeveld committed:
Cleanup comments

ronaldlangeveld committed:
Fixed onClick event handler

ronaldlangeveld committed:
Merge branch 'main' into tier-filters

ronaldlangeveld committed:
Merge branch 'main' into tier-filters

allouis commented on ghost/core/test/regression/models/model_members.test.js:
```diff
@@ -503,6 +505,74 @@ describe('Member Model', function run() {
+            const offerId = addRedemption.get('offer_id');
+
+            const members = await Member.findPage({filter: `offer_redemptions:${offerId}`});
+            should.equal(members.data.length, 1, 'Can search for members with offer redemptions');
```
I think we should check that the specific member you added in this test is the one that is returned, maybe update the email address to something like `offerredemptionfilteringtest@whatevber.comsdlkcjkl`

ronaldlangeveld replied:
Done 😊

allouis reviewed (approved):
Looks good from backend side - one comment about improving test a little

ronaldlangeveld committed:
Improved members model test

refs https://github.com/TryGhost/Team/issues/2011 https://github.com/TryGhost/Ghost/pull/16071#discussion_r1063200763

- Improved the redemptions filter test to be more specific.

kevinansfield commented on ghost/admin/app/components/gh-members-segment-select.js:
```diff
@@ -120,6 +120,26 @@ export default class GhMembersSegmentSelect extends Component {
 
+        const offers = yield this.store.query('offer', {limit: 'all'});
+
+        if (offers.length > 0 && !this.args.hideOffers) {
```
Can the `if (!this.args.hideOffers)` conditional be used to avoid waiting on the API call when it's not needed?

kevinansfield commented on ghost/admin/app/components/gh-member-settings-form.js:
```diff
@@ -23,6 +23,7 @@ export default class extends Component {
     @tracked showMemberTierModal = false;
     @tracked tiersList;
     @tracked newslettersList;
+    @tracked offersList;
```
Where is this used?

kevinansfield commented on ghost/admin/app/templates/offers.hbs:
```diff
@@ -92,9 +92,9 @@
                                 {{/if}}
                             </LinkTo>
-                            <LinkTo @route="offer" @model={{offer}} class="gh-list-data">
+                            <a href="#" {{on "click" (fn this.offerMembers offer.id)}}  class="gh-list-data pointer">
```
This can stay as a `<LinkTo>` so the behaviour is kept in the template and easily visible, query params can be set like this:

```
<LinkTo @route="members" @query={{hash filter=(concat "offer_redemptions:" offer.id)}} class="gh-list-data">
```

kevinansfield reviewed (changes requested):
A few inline comments on small things plus two other areas that need addressing:

1. the `<DefaultRecipientsSelect>` component was updated to include offers but the associated `<GhMembersRecipientSelect>` component wasn't so it's possible to set a default recipient segment that isn't selectable when publishing (these components should really be refactored to share behaviour if the design allows but that's a different task)
2. there are no Admin tests 😬

ronaldlangeveld committed:
Simplified <LinkTo> offers

refs https://github.com/TryGhost/Team/issues/2011 https://github.com/TryGhost/Ghost/pull/16071#discussion_r1063288369

ronaldlangeveld committed:
Removed unusued offerList tracker variable

refs https://github.com/TryGhost/Team/issues/2011 https://github.com/TryGhost/Ghost/pull/16071#discussion_r1063277452

- leftover from code no longer in use.

ronaldlangeveld committed:
Added offers testing to filter tests

refs https://github.com/TryGhost/Team/issues/2011

- Started adding offers redemptions testing to filter tests.
- At it tests to ensure offers will be selectable in the multi select
  dropdown in the filtering menu.
- We still need to find a way add redemptions to members (eg made
  payment using an offer) during testing as we're doing in the backend
tests.

ronaldlangeveld committed:
Merge branch 'main' into tier-filters

ronaldlangeveld committed:
Added offer redemptions members testing

ref https://github.com/TryGhost/Team/issues/2011 https://github.com/TryGhost/Ghost/pull/16071#pullrequestreview-1238582587

- Added testing for filtering members with redeemed offers and can set
  filter by path

ronaldlangeveld committed:
Cleaned up comments

no issue

ronaldlangeveld committed:
Merge branch 'main' into tier-filters

ronaldlangeveld committed:
Fixed linting

no issue
